### PR TITLE
SUPERSEDED: Iso/Interrupt streaming changes to usb_testutils

### PR DESCRIPTION
**Pull request by alees24**
This PR extends the streaming code in usb_testutils and the DPI model to handle Interrupt Transfers and Isochronous Transfers.

- Support for different Transfer Types to endpoints is introduced in usb_testutils.c/h. To keep the changes minimal, it is unfortunately necessary to have information duplicated in the _controlep.c/h code which describes the configuration(s) to the host and in the usb_testutils.c/h code which manages the endpoints; at some point these should be reconciled.
- Streaming code to support Isochronous streams is incorporated because of the potential for packet loss with Isochronous streams. The existing streaming code is modified to supply a signature at the start of each packet and not just at the start of the data stream. All data is still checked at both ends of the connection.

This is the first part of #18305 and stands alone, implementing the device end of the required functionality.

### CODE REVIEW

**Comment by alees24** [comment]
> Thanks for making these changes, I had a quick look through the code and nothing jumped out to me. Do you have any instructions for me to test this code?

This PR requires the changes in #18420 to be present in the DPI model (the other end of the USB connection), and then the two top-level tests added in #18305 exploit the new functionality. With all of those change together, the following command will build and run a Verilated top-level test:
```
bazel test --test_output=streamed --test_tag_filters=verilator //sw/device/tests:usbdev_mixed_test
```
It does require Verilator and a top-level chip sim at present, though, and the run time is something like 20-35 minutes, although at the top of the test_main() function in each of usbdev_stream_test, usbdev_iso_test and usbdev_mixed_test, it is possible to reduce the 'transfer_bytes' values to a smaller value in order to reduce the test duration.

File: sw/device/lib/testing/usb_testutils_streams.c [hunk 1]
```diff
@@ -275,6 +437,35 @@ static status_t rx_show(void *stream_v, dif_usbdev_rx_packet_info_t packet_info,
...
+  // we must ensure that there are buffers available for reception, and we
+  // do not want any endpoint to starve another
+  for (unsigned s = 0U; s < nstreams; s++) {
+    // This is slightly overspending the available buffers, leaving the
+    //   endpoints to vie for the final few buffers, so it's important that
+    //   we limit the total number of buffers across all endpoints too
+    unsigned ep = ctx->streams[s].tx_ep;
+    ctx->tx_bufs_queued[ep] = 0U;
+    ctx->tx_bufs_limit[ep] =
+        (USBUTILS_STREAMS_TXBUF_MAX + nstreams - 1) / nstreams;
```
Comment (a-will) [inline sw/device/lib/testing/usb_testutils_streams.c:459]: Does this oversubscribing of buffers enter the potential for random failures? (No analysis done on my part; this checking was merely prodded by the comments)

**Review by a-will — APPROVED** [review]
LGTM, with one question.

File: sw/device/lib/testing/usb_testutils_streams.c (hunk 1 above)
Comment (alees24) [reply sw/device/lib/testing/usb_testutils_streams.c:459]: No, the code in usb_testutils_stream_service() will only attempt to claim a 'send buffer' if the budgets would not be exceeded, and will then only proceed if it successfully claims a buffer from the DIF. Provided that all streams are regularly serviced it balances out.

`  if (tx_more &&                              // More bytes to transfer?
      nqueued < ctx->tx_bufs_limit[tx_ep] &&  // Endpoint allowed buffer?
      ctx->tx_queued_total <
          USBUTILS_STREAMS_TXBUF_MAX) {  // Total buffers not exceeded?`

Comment (a-will) [reply sw/device/lib/testing/usb_testutils_streams.c:459]: Thanks!

**Review by engdoreis — APPROVED** [review]

**Comment by marnovandermaas** [comment]
Closing this pull request because it has been superseded.

**Pull request closed without merging**